client/device/cloudftd/cloudftdonboarding/create.go

```diff
@@ -0,0 +1,136 @@
+	"time"
+)
+
+type CreateInput struct {
+	FtdUid string
+}
+
+func NewCreateInput(ftdId string) CreateInput {
+	return CreateInput{
+		FtdUid: ftdId,
+	}
+}
+
+type CreateOutput = fmcconfig.CreateDeviceRecordOutput
+
+func Create(ctx context.Context, client http.Client, createInp CreateInput) (*CreateOutput, error) {
+
+	client.Logger.Println("creating cloud ftd onboarding")
+
+	// 1. read fmc domain uid
+	client.Logger.Println("retrieving fmc domain uid")
+
+	// 1.1 read fmc
+	fmcRes, err := cloudfmc.Read(ctx, client, cloudfmc.NewReadInput())
+	if err != nil {
+		return nil, err
+	}
+	// 1.2 read fmc domain info in its specific/appliance device
+	readFmcDomainRes, err := fmcplatform.ReadFmcDomainInfo(ctx, client, fmcplatform.NewReadDomainInfo(fmcRes.Host))
+	if err != nil {
+		return nil, err
+	}
+	if len(readFmcDomainRes.Items) == 0 {
+		return nil, fmt.Errorf("failed to read fmc domain uid, fmc domain info not found")
+	}
+	fmcDomainUid := readFmcDomainRes.Items[0].Uuid
+
+	// 2. get a system token for creating FTD device record in FMC
+	// CDO token does not work, we will get a 405 method not allowed if we do that
+	client.Logger.Println("getting a system token for creating FTD device record in FMC")
+
+	// 2.1 get tenant context => tenant uid
+	readOutp, err := user.GetTokenInfo(ctx, client, user.NewGetTokenInfoInput())
+	if err != nil {
+		return nil, err
+	}
+	tenantUid := readOutp.UserAuthentication.Details.TenantUid
+	// 2.2 get a system token with scope of tenant uid
+	createTokenOutp, err := user.CreateSystemToken(ctx, client, user.NewCreateSystemTokenInput(tenantUid))
+	if err != nil {
+		return nil, err
+	}
+
+	// 3. post device record
+	client.Logger.Println("creating FTD device record in FMC")
+
+	// 3.1 read ftd metadata
+	readFtdOutp, err := cloudftd.ReadByUid(ctx, client, cloudftd.NewReadByUidInput(createInp.FtdUid))
+	if err != nil {
+		return nil, err
+	}
+	// 3.2 create ftd device
+	createDeviceInp := fmcconfig.NewCreateDeviceRecordInputBuilder().
+		Type("Device").
+		NatId(readFtdOutp.Metadata.NatID).
+		Name(readFtdOutp.Name).
+		AccessPolicyUid(readFtdOutp.Metadata.AccessPolicyUid).
+		LicenseCaps(readFtdOutp.Metadata.LicenseCaps).
+		PerformanceTier(readFtdOutp.Metadata.PerformanceTier).
+		RegKey(readFtdOutp.Metadata.RegKey).
+		FmcDomainUid(fmcDomainUid).
+		FmcHostname(fmcRes.Host).
+		SystemApiToken(createTokenOutp.AccessToken).
+		Build()
+	var createOutp fmcconfig.CreateDeviceRecordOutput
+	err = retry.Do(
+		fmcconfig.UntilCreateDeviceRecordSuccess(ctx, client, createDeviceInp, &createOutp),
+		retry.NewOptionsBuilder().
+			Retries(-1).
+			Delay(3*time.Second).
+			Timeout(1*time.Hour). // it can take 15-20 minutes for FTD to come up + 10 minutes to create device record
+			Logger(client.Logger).
+			EarlyExitOnError(false).
+			Build(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// 4. trigger FTD onboarding state machine
+	client.Logger.Println("re-triggering FTD onboarding state machine")
```

Comment:
This comment and println are slightly misleading. You're triggering the FTDC registration state machine, aren't you? I'd call it that.

Reply:
fixed 👍